### PR TITLE
[4.3] Local adapter thumbnails (PR 36552 clone)

### PR DESCRIPTION
**Pull request by obuisard**
### Summary of Changes

This Pull Request is for issue https://github.com/joomla/joomla-cms/issues/36533 [Solves no 2 of the 3 reported problems, the thumbs] and is a rewrite for https://github.com/joomla/joomla-cms/pull/36552 created by @dgrammatiko.

Please refer to the original PR if you want to follow the discussions about the implementation.

The general idea is that image thumbnails are created in the Media Manager and those thumbnails are cached, increasing performance especially when a large amount of images is processed.

The thumbnail creation is 'off' by default.

### Testing Instructions

Go to the _FileSystem - Local_ plugin and set 'Create thumbnails' to 'Yes'  on the 'images' folder.
 
### Actual result BEFORE applying this Pull Request

The media manager loads full size images.

### Expected result AFTER applying this Pull Request

The media manager loads a thumbnail image (max width or height of 200px).

### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: https://docs.joomla.org/Chunk4x:Extensions_Plugin_Manager_Edit_FileSystem_Group
- [ ] No documentation changes for docs.joomla.org needed

- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/59
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by obuisard** [comment]
tests/Codeception/api/com_media/MediaCest.php needs to be updated and fails at this point.

**Comment by dgrammatiko** [comment]
Hey, @obuisard I can't reply on #38722 as Nik has blocked me so I'll write my comment here:
- when I asked you to resurrect this PR I meant to fork it and patch the things that other were asking me to change in the original PR and also fix the failing tests
- - https://github.com/joomla/joomla-cms/pull/36552#pullrequestreview-854742766
- - https://github.com/joomla/joomla-cms/pull/36552#pullrequestreview-854741791
- - https://github.com/joomla/joomla-cms/pull/36552#discussion_r786233867
- About the other PR: the approach on that PR is wrong in the sense that you need a data structure that holds all related info per supported media file (mime type, has preview, allows edit, etc). If the project ends up with different inputs for each of all these *required* data you end up with countless inputs which makes the app really hard to configure and finally use it to its full extend. Mind that since Media manager supports audio and video and other types at some point you might want to provide edit for those (I had a trim plugin for audio and video). Also one more thing: uploads should only allowed if there is a known and matching mime type, the idea that in 2022 people rely on extensions for uploads is a huge security problem. A simple composer require is all it takes: https://packagist.org/packages/league/mime-type-detection to start making things right...

**Comment by obuisard** [comment]
Thank you Dimitris @dgrammatiko. I did address the first review, fixed formatting and tested the code successfully. 

As far as the tests are concerned, a solution would be to remove testUpdateFileWithoutAdapter all together, but I would appreciate someone's expertise on this.

**Comment by dgrammatiko** [comment]
> a solution would be to remove testUpdateFileWithoutAdapter all together,

@laoneo could you advise here?

**Comment by laoneo** [comment]
Removing the test is a bad idea. Better to set the default to 0, so no thumbs will be generated and the admin has to activate it. If it is on by default, then after the update all thumbnails are generated without notice, which can lead to resource issues on large sites.

**Comment by obuisard** [comment]
I have modified the code to make the thumbnail creation 0 by default. It does solve issues with tests. Thank you @laoneo for the suggestion.

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 53bc63a377dfc6b2bb4ddcb040ac284aa49c6e07<br /><br />Everything looks good after applying the PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38805">issues.joomla.org/tracker/joomla-cms/38805</a>.</sub>

**Comment by N6REJ** [comment]
RTC

**Comment by obuisard** [comment]
First comment on this PR was edited by the bot, removing information.
Thank you to @dgrammatiko for this work, it's a big step into improving the performance of the media manager.
For information, turning on thumbnails can be found in the Filesystem 'Local' plugin.

image

File: libraries/src/Image/Image.php [hunk 1]
```diff
@@ -366,6 +372,24 @@ public function createThumbs($thumbSizes, $creationMethod = self::SCALE_INSIDE,
...
+     * @param   integer  $creationMethod   1-3 resize $scaleMethod | 4 create cropping
+     * @param   string   $thumbsFolder     destination thumbs folder. null generates a thumbs folder in the image folder
+     *
+     * @return  array
+     *
+     * @since   2.5.0
+     * @throws  \LogicException
+     * @throws  \InvalidArgumentException
+     */
+    public function createThumbs($thumbSizes, $creationMethod = self::SCALE_INSIDE, $thumbsFolder = null)
```
Comment (laoneo) [inline libraries/src/Image/Image.php:390]: I forgot, here you need to add a deprecated statement and make a pr for the manual.

File: plugins/filesystem/local/src/Adapter/LocalAdapter.php [hunk 2]
```diff
@@ -267,22 +318,31 @@ public function updateFile(string $name, string $path, $data)
      * @since   4.0.0
      * @throws  \Exception
      */
-    public function delete(string $path)
+    public function delete(string $path): void
```
Comment (laoneo) [inline plugins/filesystem/local/src/Adapter/LocalAdapter.php]:
```suggestion
    public function delete(string $path)
```
Just to be on the safe side, when somebody extends the LocalAdapter delete function.

File: plugins/filesystem/local/src/Adapter/LocalAdapter.php [hunk 3]
```diff
@@ -814,4 +873,89 @@ private function getLocalPath(string $path): string
...
+     * Throws an InvalidPathException if the path is invalid.
+     *
+     * @param   string  $path  The path
+     *
+     * @return  array
+     *
+     * @since   __DEPLOY_VERSION__
+     * @throws  InvalidPathException
+     */
+    private function getLocalThumbPaths(string $path): array
```
Comment (laoneo) [inline plugins/filesystem/local/src/Adapter/LocalAdapter.php]: Can this be renamed to `getLocalThumbnailPaths` so it is better understandable and more inline with the change in the image class?

File: plugins/filesystem/local/src/Adapter/LocalAdapter.php [hunk 4]
```diff
@@ -814,4 +873,89 @@ private function getLocalPath(string $path): string
...
+     * Returns the path for the thumbnail of the given image.
+     * If the thumbnail does not exist, it will be created.
+     *
+     * @param   string  $path  The path of the image
+     *
+     * @return  string
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    private function getThumb(string $path): string
```
Comment (laoneo) [inline plugins/filesystem/local/src/Adapter/LocalAdapter.php]: Can this be renamed to `getThumbnail` so it is better understandable and more inline with the change in the image class?

File: plugins/filesystem/local/src/Adapter/LocalAdapter.php [hunk 5]
```diff
@@ -814,4 +873,89 @@ private function getLocalPath(string $path): string
...
+     * Create a thumbnail of the given image.
+     *
+     * @param   string  $path       The path of the image
+     * @param   string  $thumbPath  The path of the thumbnail
+     *
+     * @return  boolean
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    private function createThumb(string $path, string $thumbPath): bool
```
Comment (laoneo) [inline plugins/filesystem/local/src/Adapter/LocalAdapter.php]: Can this be renamed to `createThumbnail` so it is better understandable and more inline with the change in the image class?

File: plugins/filesystem/local/src/Adapter/LocalAdapter.php [hunk 6]
```diff
@@ -55,22 +55,52 @@ class LocalAdapter implements AdapterInterface
...
     private $filePath = null;
 
+    /**
+     * Should the adapter create a thumbnail for the image?
+     *
+     * @var boolean
+     *
+     * @since  __DEPLOY_VERSION__
+     */
+    private $thumbs = false;
```
Comment (laoneo) [inline plugins/filesystem/local/src/Adapter/LocalAdapter.php]: Can this be renamed to `thumbnails` so it is better understandable and more inline with the change in the image class?

File: plugins/filesystem/local/src/Adapter/LocalAdapter.php [hunk 7]
```diff
@@ -55,22 +55,52 @@ class LocalAdapter implements AdapterInterface
...
+
+    /**
+     * Thumbnail dimensions in pixels, [0] = width, [1] = height
+     *
+     * @var array
+     *
+     * @since  __DEPLOY_VERSION__
+     */
+    private $thumbSize = [200, 200];
```
Comment (laoneo) [inline plugins/filesystem/local/src/Adapter/LocalAdapter.php]: Can this be renamed to `thumbnailSize` so it is better understandable and more inline with the change in the image class?

File: administrator/language/en-GB/plg_filesystem_local.ini [hunk 8]
```diff
@@ -6,5 +6,6 @@
 PLG_FILESYSTEM_LOCAL="FileSystem - Local"
 PLG_FILESYSTEM_LOCAL_DEFAULT_NAME="Local"
 PLG_FILESYSTEM_LOCAL_DIRECTORIES_DIRECTORY_LABEL="Select directories"
+PLG_FILESYSTEM_LOCAL_DIRECTORIES_DIRECTORY_THUMBS_LABEL="Create thumbnails"
```
Comment (laoneo) [inline administrator/language/en-GB/plg_filesystem_local.ini]: Must here no every word start with upper case? And please write here also PLG_FILESYSTEM_LOCAL_DIRECTORIES_DIRECTORY_THUMBNAILS_LABEL.

Comment (obuisard) [reply administrator/language/en-GB/plg_filesystem_local.ini]: Agree

File: plugins/filesystem/local/src/Adapter/LocalAdapter.php [hunk 9]
```diff
@@ -293,17 +353,17 @@ public function delete(string $path)
...
+     * - name:              The name of the file
+     * - path:              The relative path to the root
+     * - extension:         The file extension
+     * - size:              The size of the file
+     * - create_date:       The date created
+     * - modified_date:     The date modified
+     * - mime_type:         The mime type
+     * - width:             The width, when available
+     * - height:            The height, when available
+     * - thumbnail_path     The thumbnail path of file, when available
```
Comment (dgrammatiko) [inline plugins/filesystem/local/src/Adapter/LocalAdapter.php]: You CANNOT do that without breaking B/C: The `thumb_path` is used here: https://github.com/joomla/joomla-cms/blob/9cbe6cb5b204705a2a66b7b774160d78eb783a01/build/media_source/system/js/fields/joomla-media-select.w-c.es6.js#L132

**Review by dgrammatiko — CHANGES_REQUESTED** [review]

**Comment by richard67** [comment]
Back to pending because changes were made and more changes were requested.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38805">issues.joomla.org/tracker/joomla-cms/38805</a>.</sub>

File: plugins/filesystem/local/src/Adapter/LocalAdapter.php [hunk 10]
```diff
@@ -346,8 +406,7 @@ private function getPathInformation(string $path): \stdClass
                 $obj->width  = $props->width;
                 $obj->height = $props->height;
 
-                // @todo : Change this path to an actual thumbnail path
-                $obj->thumb_path = $this->getUrl($obj->path);
+                $obj->thumbnail_path = $this->thumbnails ? $this->getThumbnail($path) : $this->getUrl($obj->path);
```
Comment (dgrammatiko) [inline plugins/filesystem/local/src/Adapter/LocalAdapter.php]: Same here

File: plugins/filesystem/local/src/Adapter/LocalAdapter.php (hunk 9 above)
Comment (laoneo) [reply plugins/filesystem/local/src/Adapter/LocalAdapter.php]: Yes you are right, sorry @obuisard , I was wrong here.

File: plugins/filesystem/local/src/Adapter/LocalAdapter.php (hunk 10 above)
Comment (laoneo) [reply plugins/filesystem/local/src/Adapter/LocalAdapter.php]: Again you are right!! Thanks for checking.

Comment (obuisard) [reply plugins/filesystem/local/src/Adapter/LocalAdapter.php]: Yep! Thank you both for the review, I have checked the file again

File: plugins/filesystem/local/src/Adapter/LocalAdapter.php (hunk 9 above)
Comment (obuisard) [reply plugins/filesystem/local/src/Adapter/LocalAdapter.php]: Yes, I have figured it out while testing it. Thank you both.

File: libraries/src/Image/Image.php (hunk 1 above)
Comment (obuisard) [reply libraries/src/Image/Image.php:390]: I have added @deprecated, not sure if 5.0, 6.0 or 7.0...

Comment (laoneo) [reply libraries/src/Image/Image.php:390]: Must be 6.0. A pr similar to [this](https://github.com/joomla/Manual/pull/50) one is required as well in the manual.

**Comment by obuisard** [comment]
Troy @N6REJ and Jacob @jwaisner, would you mind re-testing this so we can merge it? Thank you so much!

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on d82abd9fed569dccc9e8772486ae20137d4fd510<br /><br />Everything retested well. No issues.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38805">issues.joomla.org/tracker/joomla-cms/38805</a>.</sub>

**Comment by N6REJ** [comment]
I have tested this item :red_circle: unsuccessfully on d82abd9fed569dccc9e8772486ae20137d4fd510<br /><br />in my test the images were already thumbnailed without the PR.  This was " The currently installed Joomla! version is "‎4.3.0-dev"‎4.3.0-dev"
with the latest composer & npm.

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38805">issues.joomla.org/tracker/joomla-cms/38805</a>.</sub>

**Comment by obuisard** [comment]
> in my test the images were already thumbnailed without the PR. 

That's strange. The thumbnails created with the PR should be 200x200.

**Comment by dgrammatiko** [comment]
@obuisard what @N6REJ is showcasing is the SIZE of the container DIV, not the size of the image. To get the size of the image open the browser tools, select the image in the elements tab and in the console type `$0.naturalWidth` (or naturalHeight for the height). False alarm...

**Comment by obuisard** [comment]
> @obuisard what @N6REJ is showcasing is the SIZE of the container DIV, not the size of the image. To get the size of the image open the browser tools, select the image in the elements tab and in the console type `$0.naturalWidth` (or naturalHeight for the height). False alarm...

Oh good catch Dimitris @dgrammatiko! @N6REJ can you double-check? Thank you!

**Comment by Quy** [comment]
Blurry quality:
38805-blurry-quality

Before PR:
38805-gif-before

After PR:
38805-gif-after

**Comment by obuisard** [comment]
Thank you, Troy @N6REJ for reporting those issues. 
The image quality and handling of transparency for GIFs is unrelated to this PR, but improvements need to be made to the image library itself. Something for another PR...

**Comment by N6REJ** [comment]
I have tested this item :white_check_mark: successfully on d82abd9fed569dccc9e8772486ae20137d4fd510<br /><br />retested as instructed.

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38805">issues.joomla.org/tracker/joomla-cms/38805</a>.</sub>

**Comment by obuisard** [comment]
A big thank you to Dimitris @dgrammatiko and the testers !

**Comment by carlitorweb** [comment]
**Documentation added** :heavy_check_mark:

**Comment by dgrammatiko** [comment]
@obuisard I haven't checked but is the thumbnails functionality bypasses the svg files? If not it needs to do so, the thumbs helper expects only files that the Image class can handle. I'll try to check the next days if I get some free time

**Comment by obuisard** [comment]
> @obuisard I haven't checked but is the thumbnails functionality bypasses the svg files? If not it needs to do so, the thumbs helper expects only files that the Image class can handle. I'll try to check the next days if I get some free time

@dgrammatiko I have checked SVG and image thumbnail creation and both work well together

**Comment by N6REJ** [comment]
it's sad that we are using GD instead of imagemagick.

**Comment by obuisard** [comment]
> it's sad that we are using GD instead of imagemagick.

I know, but GD is more widely installed than imagick. It would be great to offer the use of either one of those graphic libraries in Joomla.

**Comment by dgrammatiko** [comment]
You'll be surprised but GD (PHP 8.1+) is (median) faster than imagick. How do I know? I keep testing these for [responsive-images](https://responsive-images.dgrammatiko.dev)

**Comment by N6REJ** [comment]
@dgrammatiko faster but I thouight imagick quality was far higher?

**Pull request merged**